library/core/tests/iter/range.rs

```diff
@@ -467,3 +467,255 @@ fn test_double_ended_range() {
+
+    let mut range = start..end;
+    assert_eq!(range.advance_by(14), Err(13));
+    assert_eq!(range, end..end);
+
+    let mut range = start..end;
+    assert_eq!(range.next_back(), Some(12_04 as _));
+    assert_eq!(range, start..12_04 as _);
+
+    let mut range = start..end;
+    assert_eq!(range.nth_back(0), Some(12_04 as _));
+    assert_eq!(range, start..12_04 as _);
+
+    let mut range = start..end;
+    assert_eq!(range.nth_back(1), Some(11_04 as _));
+    assert_eq!(range, start..11_04 as _);
+
+    let mut range = start..end;
+    assert_eq!(range.nth_back(12), Some(start));
+    assert_eq!(range, start..start);
+
+    let mut range = start..end;
+    assert_eq!(range.advance_back_by(0), Ok(()));
+    assert_eq!(range, start..end);
+
+    let mut range = start..end;
+    assert_eq!(range.advance_back_by(1), Ok(()));
+    assert_eq!(range, start..12_04 as _);
+
+    let mut range = start..end;
+    assert_eq!(range.advance_back_by(12), Ok(()));
+    assert_eq!(range, start..1_04 as _);
+
+    let mut range = start..end;
+    assert_eq!(range.advance_back_by(13), Ok(()));
+    assert_eq!(range, start..start);
+
+    let mut range = start..end;
+    assert_eq!(range.advance_back_by(14), Err(13));
+    assert_eq!(range, start..start);
+}
+
+#[test]
+fn test_ptr_range_zero_sized() {
+    let start = 8 as *const ();
+    let mut range = start..start;
+    assert_eq!(range.size_hint(), (0, Some(0)));
+    assert_eq!(range.next(), None);
+    assert_eq!(range.nth(0), None);
+    assert_eq!(range.nth(1), None);
+    assert_eq!(range.advance_by(0), Ok(()));
+    assert_eq!(range.advance_by(1), Err(0));
+    assert_eq!(range.next_back(), None);
+    assert_eq!(range.nth_back(0), None);
+    assert_eq!(range.nth_back(1), None);
+    assert_eq!(range.advance_back_by(0), Ok(()));
+    assert_eq!(range.advance_back_by(1), Err(0));
+    assert_eq!(range, start..start);
+
+    let end = 11 as *const ();
+    let mut range = start..end;
+    assert_eq!(range.size_hint(), (usize::MAX, None));
+    assert_eq!(range.next(), Some(start));
+    assert_eq!(range.nth(0), Some(start));
+    assert_eq!(range.nth(1), Some(start));
+    assert_eq!(range.advance_by(0), Ok(()));
+    assert_eq!(range.advance_by(1), Ok(()));
+    assert_eq!(range.next_back(), Some(start));
+    assert_eq!(range.nth_back(0), Some(start));
+    assert_eq!(range.nth_back(1), Some(start));
+    assert_eq!(range.advance_back_by(0), Ok(()));
+    assert_eq!(range.advance_back_by(1), Ok(()));
+    assert_eq!(range, start..end);
+}
+
+#[test]
+fn test_ptr_range_reversed() {
+    let start = 2222 as *const i32;
+    let end = 1111 as *const i32;
+    let range = start..end;
+    assert!(range.is_empty());
+    assert_eq!(range.size_hint(), (0, Some(0)));
+    assert_eq!(range.clone().next(), None);
+    assert_eq!(range.clone().nth(0), None);
+    assert_eq!(range.clone().nth(1), None);
+    assert_eq!(range.clone().advance_by(0), Ok(())); // Err(0) would be equally fine
```

Comment:
It's better to not add that comment because the flexibility granted by the current `advance_by` contract turned out to be inconsistent and increases the number of cases that need to be handled, after some discussion in #77404 I have filed  #89916 to remove that again.

So `Ok(())` is the correct one if that PR gets accepted.

Reply:
Ah nice. Fixed in 4cb61510e1bdbf289e939eb3d75b2e876b133b4b. I've approved your PR.